### PR TITLE
Non-blocking autologin

### PR DESCRIPTION
**Pull request by lopuhin**
I make requests to autologin via normal scrapy requests and process responses in callbacks. Requests before login and during logout are queued and scheduled after login.

### CODE REVIEW

**Comment by lopuhin** [comment]
Hm, strange that this https://github.com/TeamHG-Memex/undercrawler/pull/46/files#diff-cd4de073722fd256d3d944b28a9c88baR95 is not covered, may be something fishy here, I was sure it must be covered...

File: undercrawler/middleware/autologin.py [hunk 1]
```diff
@@ -36,16 +37,21 @@ class AutologinMiddleware:
     - do not block event loop in login() method (instead, collect
     scheduled requests in a separate queue and make request with scrapy).
```
Comment (kmike) [inline undercrawler/middleware/autologin.py]: is the docstring still valid?

Comment (lopuhin) [inline undercrawler/middleware/autologin.py]: Nice catch, thanks! Fixed, I think now the only restriction is a single authorization domain per spider, and now it should be easier to relax if needed.

**Comment by lopuhin** [comment]
I think it's ready now @kmike !

I tried to come up with a scrapy API that would simplify this case, but did not come up with anything good. One small thing that could simplify it is making ``crawler.engine.crawl(request, spider)`` more obvious, perhaps as ``spider.crawl(request)``, but this does not change much. A more powerful thing would be to add a ``DelayRequest`` exception, so instead of adding request to ``self._queue`` it would be queued somewhere in scrapy, but then we should also have some way to say when to re-process this (original) requests, and it also must be composable, so that several middlewares can use it, so it starts looking more complicated than explicitly queueing requests in the middleware.

File: undercrawler/middleware/autologin.py [hunk 2]
```diff
@@ -55,44 +59,83 @@ def __init__(self, autologin_url, auth_cookies=None, logout_url=None,
...
+            # TODO - it should be possible to put auth cookies into the
+            # cookiejar in process_response (but also check non-splash)
+            if self.auth_cookies:
+                request.cookies = self.auth_cookies
+                autologin_meta['cookie_dict'] = {
+                    c['name']: c['value'] for c in self.auth_cookies}
+        else:
+            self._enqueue(request)
+            if self.waiting_for_login:
+                raise IgnoreRequest
```
Comment (kmike) [inline undercrawler/middleware/autologin.py]: Can we solve it without dropping the request? Downloader middleware support returning Deferreds from process_request; see e.g. https://github.com/scrapy/scrapy/blob/master/scrapy/downloadermiddlewares/robotstxt.py implementation.

Comment (lopuhin) [inline undercrawler/middleware/autologin.py]: It looks much nicer, I'll try, thanks!

Comment (lopuhin) [inline undercrawler/middleware/autologin.py]: It tool a while to get used to such style, but at the end it's better, I think. The only gotcha is that at one point the traceback was incorrect, is it worth a bug report?

Comment (kmike) [inline undercrawler/middleware/autologin.py]: Yeah, bad tracebacks worth a bug report.

Comment (lopuhin) [inline undercrawler/middleware/autologin.py]: Yep, filed https://github.com/scrapy/scrapy/issues/1948

**Pull request merged**